java-squid/src/main/java/org/sonar/java/cfg/CFG.java

```diff
@@ -0,0 +1,691 @@
+        Block switchSuccessor = currentBlock;
+        // process condition
+        currentBlock = createBlock();
+        currentBlock.terminator = switchStatementTree;
+        switches.addLast(currentBlock);
+        build(switchStatementTree.expression());
+        // process body
+        currentBlock = createBlock(switchSuccessor);
+        breakTargets.addLast(switchSuccessor);
+        if (!switchStatementTree.cases().isEmpty()) {
+          CaseGroupTree firstCase = switchStatementTree.cases().get(0);
+          for (CaseGroupTree caseGroupTree : switchStatementTree.cases()) {
+            build(caseGroupTree.body());
+            switches.getLast().successors.add(currentBlock);
+            if (caseGroupTree != firstCase) {
+              // No block predecessing the first case group.
+              currentBlock = createBlock(currentBlock);
+            }
+          }
+        }
+        breakTargets.removeLast();
+        // process condition
+        currentBlock = switches.removeLast();
+        break;
+      }
+      case BREAK_STATEMENT: {
+        BreakStatementTree breakStatementTree = (BreakStatementTree) tree;
+        if(breakStatementTree.label() == null) {
+          if (breakTargets.isEmpty()) {
+            throw new IllegalStateException("'break' statement not in loop or switch statement");
+          }
+          currentBlock = createUnconditionalJump(tree, breakTargets.getLast());
+        } else {
+          currentBlock = createUnconditionalJump(tree, null);
+          gotos.add(currentBlock);
+        }
+        break;
+      }
+      case CONTINUE_STATEMENT: {
+        ContinueStatementTree continueStatementTree = (ContinueStatementTree) tree;
+        if(continueStatementTree.label() == null) {
+          if (continueTargets.isEmpty()) {
+            throw new IllegalStateException("'break' statement not in loop or switch statement");
+          }
+          currentBlock = createUnconditionalJump(tree, continueTargets.getLast());
+        } else {
+          currentBlock = createUnconditionalJump(tree, null);
+          gotos.add(currentBlock);
+        }
+        break;
+      }
+      case WHILE_STATEMENT: {
+        WhileStatementTree s = (WhileStatementTree) tree;
+        Block falseBranch = currentBlock;
+        Block loopback = createBlock();
+        // process body
+        currentBlock = createBlock(loopback);
+        continueTargets.addLast(loopback);
+        breakTargets.addLast(falseBranch);
+        build(s.statement());
+        breakTargets.removeLast();
+        continueTargets.removeLast();
+        Block bodyBlock = currentBlock;
+        // process condition
+        currentBlock = createBranch(s, bodyBlock, falseBranch);
+        buildCondition(s.condition(), bodyBlock, falseBranch);
+        loopback.successors.add(currentBlock);
+        currentBlock = createBlock(currentBlock);
+        break;
+      }
+      case DO_STATEMENT: {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Reduce this switch case number of lines from 18 to at most 5, for example by extracting code into methods. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1151)
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Extract this nested code block into a method. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1199)
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Make this line start at column 9. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AIndentationCheck)